File: Chapter 05/Python/section_5_4_server_monitor/after/core/domain/server_monitor.py
from ...core.domain import constants
from ...core.ports.alert_port import AlertPort

class ServerMonitor:
    """THE INSIDE (The Core).
    
    This class represents the Pure Domain Logic of the system. It has been 
    'Isolated' from the infrastructure, meaning it contains zero references 
    to low-level details like Console, Twilio, or Kafka.
    """

    def __init__(self, alert_port: AlertPort):
        """Initializes the monitor with an abstract alert port.
        
        Args:
            alert_port (AlertPort): An implementation of the AlertPort contract
                (Constructor Injection). This allows the Core to remain 
                agnostic of the specific notification technology used.
        """
        self.alert_port = alert_port

    def check_temperature(self, temp: int) -> None:
        """Evaluates a temperature reading against business rules.
        
        The Core acts as the 'Boundary Keeper' here—it defines 'What' needs 
        to happen (an alert), while delegating the 'How' (the transport) 
        to the outside world.

        Args:
            temp (int): The current server temperature in degrees.
        """
        if temp > constants.HIGH_TEMP_THRESHOLD:
            # Domain logic decides that an alert is necessary
            self.alert_port.send_alert(f"Temp is {temp} degrees! Take cover!")
        else:
            print(f"[Core] Temp {temp} is normal.")

File: Chapter 05/C#/Section_5_4_Server_Monitor/After/Core/Domain/ServerMonitor.cs
using Chapter05.ServerMonitor.After.Core.Ports;

namespace Chapter05.ServerMonitor.After.Core.Domain
{
    /// <summary>
    /// THE INSIDE (The Core).
    /// This is the Pure Domain Logic. It has been 'Isolated' from the 
    /// infrastructure using the Bulkhead pattern approach for code.
    /// It contains zero references to Console, Twilio, or Kafka.
    /// </summary>
    public class ServerMonitor
    {
        private readonly IAlertPort _alertPort;

        /// <summary>
        /// Constructor Injection.
        /// We "plug in" the adapter via the constructor, allowing the 
        /// Core to remain agnostic of the specific implementation.
        /// </summary>
        public ServerMonitor(IAlertPort alertPort)
        {
            _alertPort = alertPort;
        }

        /// <summary>
        /// Evaluates temperature against domain constants.
        /// </summary>
        public void CheckTemperature(int temp)
        {
            // We use the Global Constant to ensure architectural consistency.
            if (temp > Constants.HIGH_TEMP_THRESHOLD) 
            {
                // The Core just calls the Port. 
                // It acts as the 'Boundary Keeper,' defining 'What' needs to 
                // happen, while leaving the 'How' to the outside world.
                _alertPort.SendAlert($"Temp is {temp} degrees! Take cover!");
            }
            else
            {
                // Note: In a strictly pure Hexagon, even this Console call 
                // would be moved to an Output Port, but we keep it here for simplicity.
                Console.WriteLine($"[Core] Temp {temp} is normal.");
            }
        }
    }
}